torch/fx/experimental/Partitioner.py

```diff
@@ -63,120 +34,367 @@ class Partition:
+            if node.op == 'output':
+                break
+            if node.op in {'placeholder', 'get_attr'}:
+                continue
+            if node.target == operator.__getitem__:
+                continue
+            input_nodes : Dict[Node, None] = {}
+            map_arg(node.args, lambda n: input_nodes.setdefault(n))
+            map_arg(node.kwargs, lambda n: input_nodes.setdefault(n))
+            # When a node has two or more output nodes,
+            # it outputs its result to 'getitem' nodes.
+            # Those 'getitem' nodes are the output node for this node.
+            # Otherwise, the output node is this node itself.
+            if len(node.users) > 1:
+                output_nodes = list(node.users)
+            else:
+                output_nodes = [node]
+            dag.create_node(node, list(input_nodes), output_nodes)
         return dag
 
     def create_partition(self) -> Partition:
         """Create a partition and append it to self.partitions."""
         partition_id = len(self.partitions)
         assert isinstance(self.graph_module, GraphModule)
         partition = Partition(partition_id, self.graph_module)
-        self.partitions.append(partition)
+        self.partitions.add(partition)
         return partition
+
+    def combine_partitions_based_on_size(self, available_mem_bytes) -> None:
+        """Combining small partitions together to keep as less partitions as possible.
+           Here is an example of the algorithm to do this:
+           Assume some partitions, we first sort them based on partiiton used memory size.
+           [(partition_4, 1), (partition_3, 1), (partition_2, 2), (partition_1, 7), (partition_0, 9)]
+           The available memory is 10.
+           step 1: self.find_partition_to_combine_based_on_size()
+           First, look the smallest partition, partition_4: 10 - 1 = 9
+           It means any partition has a used memory equal or less than 9 could combine this partition
+           We go from the largest and selection partition_0.
+           Then we will check if there is a circular dependency between partitions using self.check_partition_dependency
+           If there is no cycle, partition_9 will be the combined partition of partition_0 and partition_4
+           As a result, partition_9 has a size of 10
+           Then, repeat step 1 for the rest.
+        """
+        find_combination = True
+        while find_combination:
+            sorted_partitions = sorted(self.partition_to_used_mem_bytes.items(), key=lambda item: item[1])
+            smallest_partition = sorted_partitions[0][0]
+            left_mem = available_mem_bytes - self.partition_to_used_mem_bytes[smallest_partition]
+            find_combination = self.find_partition_to_combine_based_on_size(sorted_partitions, left_mem)
+        return
+
+    def find_partition_to_combine_based_on_size(self, sorted_partitions: List[Tuple[Partition, int]], left_mem: int) -> bool:
```

Comment:
Note: O(N^2) complexity

Since this can run N iterations that call `check_partition_dependency`, which has complexity O(N)